runtime/tests/org.eclipse.core.tests.runtime/src/org/eclipse/core/tests/runtime/jobs/JobGroupTest.java

```diff
@@ -1293,6 +1310,76 @@ public IStatus run(IProgressMonitor monitor) {
 		assertTrue("2.0", numShouldCancelCalled[0] < NUM_JOBS_LIMIT + 10);
 	}
 
+	/**
+	 * Tested scenario: - 2 jobs are started in parallel and the first one returns a
+	 * CANCEL_STATUS while the second one is still running. This lets
+	 * <code>JobGroup.shouldCancel</code> return <code>true</code> which triggers
+	 * the cancelation of any running job.
+	 *
+	 * Expected result: The second job is explicitly canceled
+	 *
+	 */
+	public void testShouldCancel_5() {
+
+		// the job group allows for 2 threads so both jobs will be started in
+		// parallel
+		final JobGroup jobGroup = new JobGroup("JobGroup", 2, 2) {
+			@Override
+			protected boolean shouldCancel(IStatus lastCompletedJobResult, int numberOfFailedJobs,
+					int numberOfCanceledJobs) {
+				// cancel the group as soon as one job in it returns the CANCEL_STATUS
+				return numberOfCanceledJobs > 0;
+			}
+		};
+
+		final TestBarrier2 returnsCancelStatusBarrier = new TestBarrier2();
+		Job returnsCancelStatusJob = Job.create("The one that cancels", __ -> {
+			returnsCancelStatusBarrier.setStatus(TestBarrier2.STATUS_WAIT_FOR_RUN);
+			returnsCancelStatusBarrier.waitForStatus(TestBarrier2.STATUS_RUNNING);
+
+			return Status.CANCEL_STATUS;
+		});
+
+		final TestBarrier2 canceledJobBarrier = new TestBarrier2();
+		AtomicBoolean jobWasCanceled = new AtomicBoolean();
+		Job canceledJob = new Job("The one that is explicitly canceled") {
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				canceledJobBarrier.setStatus(TestBarrier2.STATUS_WAIT_FOR_RUN);
+				canceledJobBarrier.waitForStatus(TestBarrier2.STATUS_RUNNING);
+
+				// give it time: this Job needs to be running at the time the other job finishes
+				canceledJobBarrier.waitForStatus(TestBarrier2.STATUS_DONE);
+				return Status.OK_STATUS;
+			}
+
+			@Override
+			protected void canceling() {
+				jobWasCanceled.set(true);
+				canceledJobBarrier.setStatus(TestBarrier2.STATUS_DONE);
+			}
+		};
+
+		returnsCancelStatusJob.setJobGroup(jobGroup);
+		returnsCancelStatusJob.schedule();
+
+		canceledJob.setJobGroup(jobGroup);
+		canceledJob.schedule();
+
+		// Wait until both jobs are ready to run and let them start running
+		returnsCancelStatusBarrier.waitForStatus(TestBarrier2.STATUS_WAIT_FOR_RUN);
+		canceledJobBarrier.waitForStatus(TestBarrier2.STATUS_WAIT_FOR_RUN);
+
+		returnsCancelStatusBarrier.setStatus(TestBarrier2.STATUS_RUNNING);
+		canceledJobBarrier.setStatus(TestBarrier2.STATUS_RUNNING);
```

Comment:
Having two barriers is good to clarify to which job its state blongs. Still, I find the current barrier handling quite "complex". The intended semantics is as follows: Job 1 proceeds only when Job 2 started and vice versa, i.e., each job should set its barrier to "running" once started and wait for the other to be "running" before proceeding. In particular, the barrier is not about having a central mediator (the main thread) but two jobs waiting for specific state changes of the other. So I would still propose to remove the barrier logic from the main thread (execept for the wait for the final execution of the jobs) and leave the coordination up to the jobs.

Reply:
Thank you, I used 2 barriers and removed the orchestration from the main thread, as requested.